File: old/2020/10_adapter_array/aoc10.jl
# Adapter Array
#
# Advent of Code 2020, day 10
# Solution by Geir Arne Hjelle, 2020-12-10

module AOC10

using Pipe
using StatsBase: countmap

FORBIDDEN = Dict(0 => Set(), 1 => Set(), 2 => Set(), 3 => Set(), 4 => Set(["000"]))


# Count number of combinations for a given run length
function count_combinations(run_length)
    2^(run_length - 1) - length(invalid_combinations(run_length))
end

# Which combinations are invalid because of jumps >= 3
function invalid_combinations(run_length)
    if !(run_length in keys(FORBIDDEN))
        previous = invalid_combinations(run_length - 1)
        FORBIDDEN[run_length] = @pipe [
            Set("0$(p)" for p in previous),
            Set("1$(p)" for p in previous),
            Set("$(p)0" for p in previous),
            Set("$(p)1" for p in previous),
        ] |> foldl(union, _)
    end
    FORBIDDEN[run_length]
end

# Solve the problem for one file
function solve(input)
    # Parse input
    adapters = @pipe split(input, "\n") .|> parse(Int32, _) |> sort

    # Add charger (0) and adapter (max + 3)
    jumps = vcat(0, adapters, maximum(adapters) + 3) |> diff

    # Part 1
    part_1 = @pipe (
        jumps
        |> countmap  # Count occurences of each unique element
        |> [_[1], _[3]]  # Pick out 1 and 3
        |> prod
    )

    # Part 2
    part_2 = @pipe (
        vcat(3, jumps)  # Add 3 in front to control where runs of 1 start
        |> diff
        |> findall(x -> x != 0, _)  # Indices of starts of runs
        |> diff
        |> _[1:2:end]  # Pick out runs of 1s
        |> map(count_combinations, _)
        |> prod  # Total number of combinations
    )

    part_1, part_2
end


# Solve the problem for one file
function solve_file(file_path)
    println("\n$(file_path)")
    input = open(file_path) do fid
        read(fid, String) |> strip
    end
    input .|> solve
end

# Solve the problem for each file
[a for a in ARGS if a[1] != '-'] .|> solve_file .|> s -> join(s, "\n") |> println

end  # module

File: old/2020/10_adapter_array/aoc10.py
"""Adapter Array

Advent of Code 2020, day 10
Solution by Geir Arne Hjelle, 2020-12-10
"""
# Standard library imports
import pathlib
import sys

# Third party imports
import numpy as np

debug = print if "--debug" in sys.argv else lambda *_: None

FORBIDDEN = {0: {}, 1: {}, 2: {}, 3: {}, 4: {"000"}}


def count_jumps(jumps):
    """Count number of 1- and 3-jumps"""
    size, count = np.unique(jumps, return_counts=True)
    jump_counts = {s: c for s, c in zip(size, count)}
    return jump_counts[1], jump_counts[3]


def find_runs(jumps):
    """Find runs of 1 jumps"""
    # Add a 3 at front to be sure the run starts with 3s
    return np.diff(np.where(np.diff([3] + list(jumps)))[0])[::2]


def forbidden_combos(run_length):
    """Find forbidden combinations based on run lengths"""
    if run_length not in FORBIDDEN:
        previous = forbidden_combos(run_length - 1)
        FORBIDDEN[run_length] = set.union(
            *[{f"0{c}", f"1{c}", f"{c}0", f"{c}1"} for c in previous]
        )

    return FORBIDDEN[run_length]


def runs_to_combinations(run_lengths):
    """Count possible combinations based on run lengths"""
    return [2 ** (rl - 1) - len(forbidden_combos(rl)) for rl in run_lengths]


def main(args):
    """Solve the problem for all file paths"""
    for file_path in [pathlib.Path(p) for p in args if not p.startswith("-")]:
        solve(file_path)


def solve(file_path):
    """Solve the problem for one file path"""
    print(f"\n{file_path}:")
    numbers = sorted(int(n) for n in file_path.open())
    joltage_jumps = np.diff([0] + numbers + [max(numbers) + 3])

    # Part 1
    part_1 = np.prod(count_jumps(joltage_jumps))
    print(f"The product of 1- and 3-jumps is {part_1}")

    # Part 2
    part_2 = np.prod(runs_to_combinations(find_runs(joltage_jumps)))
    print(f"There are {part_2} distinct arrangements")


if __name__ == "__main__":
    main(sys.argv[1:])
